### DISCUSSION TITLE
ESP32 DAC : Sine Wave Generator

### DISCUSSION THREAD

**Discussion by scruss** [micropython/micropython#15446] [ESP32]
I've been trying to generate simple sine-wave tones using the ESP32's DAC. This article has much of the code to do it: [ESP32 In MicroPython: Using Hardware Registers](https://www.i-programmer.info/programming/148-hardware/16391-esp32-in-micropython-using-hardware-registers.html?start=2).

Unfortunately, this code only allows very crude steps between frequencies, resulting in truly awful approximation of tunes (for instance: https://xoxo.zone/@scruss/112759584284282346 — beware, it sounds terrible). What I *think* I'm looking for is the `RTC_CNTL_CK8M_DIV_SEL` field in the register `RTC_CNTL_CLK_CONF_REG`, but I'm not sure where that is or how it's applied. Can anyone help, please?

I know I'll likely need to iterate values for this divider along with the frequency step, as suggested in this Arduino code: [DacTone/DacTone.cpp · bhall66/DacTone](https://github.com/bhall66/DacTone/blob/e1f15d933597d484e27785193d50310a56d39352/DacTone.cpp#L170). But that will come after learning a bit more about mem32 things on esp32

**Comment by ned-pcs** [comment]
You shouldn't have to mess with the RTC clocks to make waveforms if you have a DAC.
Use a [numerically controlled oscillator](https://www.wikiwand.com/en/Numerically_controlled_oscillator); then you can get very good frequency resolution.
